### PR TITLE
[Product discount] Percentage discount picker

### PR DESCRIPTION
**Pull request by samiuelson**
### Percentage discount picker
⚠️ Should be reviewed and merged after https://github.com/woocommerce/woocommerce-android/pull/9392 because this branch originates from `9352-product-discounts-full-flow` (pending merge to trunk).

### Description

This PR adds possibility to enter discount amount as a percentage of the item price.

### Testing instructions

1. Create order containing items.
2. Click on the item, and then on product details screen click "+ Add discount"
3. Select a percentage discount type and enter the amount
4. Verify the correct amount is applied to the order

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/4527432/41774979-d588-4038-978e-e540494da616

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/discount/OrderCreateEditProductDiscountViewModel.kt [hunk 1]
```diff
@@ -1,11 +1,211 @@
...
+        if (this > BigDecimal.ZERO) this else null
+    }
+
+    val viewState: StateFlow<ViewState> =
+        combine(discount, discountType) { discount, type ->
+            ViewState(
+                currency = currency,
+                discountAmount = discount,
+                discountValidationState = checkDiscountValidationState(discount, type),
+                isRemoveButtonVisible = getRemoveButtonVisibility(),
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/discount/OrderCreateEditProductDiscountViewModel.kt:66]: ❓ Can we write a test to verify the behavior of the button's visibility?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/discount/OrderCreateEditProductDiscountScreen.kt [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                onClick = onDoneClicked,
+                enabled = isDoneButtonEnabled,
+                text = stringResource(id = R.string.done)
+            )
+        },
+    )
+}
+
+@Composable
+private fun Switch(
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/discount/OrderCreateEditProductDiscountScreen.kt:146]: Nitpick: ❓ Can we also have a separate preview for Switch Composable? It makes it easy to visualize individual components.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/details/OrderCreateEditProductDetailsViewModel.kt [hunk 3]
```diff
@@ -95,25 +96,27 @@ class OrderCreateEditProductDetailsViewModel @Inject constructor(
...
+        triggerEvent(NavigationTarget.DiscountEdit(item.value, currency))
     }
 
     fun onCloseClicked() {
-        if (args.item != this.item.value) {
-            triggerEvent(ExitWithResult(ProductDetailsEditResult.ProductDetailsEdited(this.item.value)))
-        } else {
-            triggerEvent(MultiLiveEvent.Event.Exit)
-        }
+        triggerEvent(ExitWithResult(ProductDetailsEditResult.ProductDetailsEdited(this.item.value)))
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/details/OrderCreateEditProductDetailsViewModel.kt:107]: ❓ I struggled for a few minutes wondering why the discount wasn't being saved. Then when I browsed the code I realized that the discounts will be applied only on the "cross" button click and will be discarded on simply navigating back. 

Can we somehow make this more clear for users in the UI? Pressing back won't save the discount vs clicking the "cross" mark saves it.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TypedTextField.kt [hunk 4]
```diff
@@ -177,6 +178,34 @@ class NullableBigDecimalTextFieldValueMapper(
     }
 }
 
+class NullableCurrencyTextFieldValueMapper(
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TypedTextField.kt:181]: ❓ Can we write tests for this class? Seems like it contains quite a bit of logic.

For example:

```
@Test
    fun `when user enters accepted characters, then return the entered text`() {
        val sut = NullableCurrencyTextFieldValueMapper(".", 2)

        assertThat(".33").isEqualTo(sut.transformText("123", ".33"))
    }
```

```
@Test
    fun `given no of decimals allowed is 2, when the user enters more than 2 decimals, then return the old string`() {
        val sut = NullableCurrencyTextFieldValueMapper(".", 2)

        assertThat("123").isEqualTo(sut.transformText("123", ".334"))
    }
```

Likewise for other logic. WDYT?

**Review by AnirudhBhat — APPROVED** [review]
Thanks for your work on this @samiuelson. The feature works as expected. I've left a few minor comments which are non-blocking.

One observation: It seems like the discount applied isn't shown as a separate row in the Order creation screen. It's hard to see what were the product's original price, discount price, and product price after applying the discount.

For example, The below image shows a product for which I've applied a discount of 90%. But from the image below, It's hard to see what's what.

Screenshot_20230713-140213

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/details/OrderCreateEditProductDetailsViewModel.kt (hunk 3 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/details/OrderCreateEditProductDetailsViewModel.kt:107]: Good catch, I'll improve this in the next PR 👍🏼

**Comment by samiuelson** [comment]
> Thanks for your work on this @samiuelson. The feature works as expected. I've left a few minor comments which are non-blocking.
> 
> One observation: It seems like the discount applied isn't shown as a separate row in the Order creation screen. It's hard to see what were the product's original price, discount price, and product price after applying the discount.
> 
> For example, The below image shows a product for which I've applied a discount of 90%. But from the image below, It's hard to see what's what.
> 

Thanks for the review and feedback. As for discount amount visibility in order creation screen, we decided to add an additional row to the Payments section displaying the total discount amount and show the item-specific discount amount in the items list below the item's total price. More context here: p1689084213840619/1689083832.564069-slack-C025A8VV728. 
This is also on the [To-Do](https://github.com/woocommerce/woocommerce-android/issues/9352) and will be added in the next PR.

I'll also improve back press handling in Product details 👍🏼

**Pull request merged**